src/modules/leaderboard.ts

```diff
@@ -0,0 +1,182 @@
+
+	const lines = [];
+
+	for (const [uid, cs] of chars) {
+		if (lines.length > MAX_PEOPLE) break;
+		lines.push(
+			`\`#${(lines.length + 1)
+				.toString()
+				.padStart(
+					Math.ceil(Math.log10(MAX_PEOPLE)),
+					"0"
+				)}\` <@${uid}> (${cs})`
+		);
+	}
+
+	lines.unshift(
+		"**__LEADERBOARD__** *(by character count)* `[" +
+			period.toUpperCase() +
+			"]`"
+	);
+
+	const sent = await sendChannel.send({
+		content: lines.join("\n"),
+		allowedMentions: { parse: [] },
+	});
+
+	return [channels, msgs, failed, sent.id];
+}
+
+export function provideCommands(): CommandDescriptor[] {
+	const cmd = new Builders.SlashCommandBuilder()
+		.setName("leaderboard")
+		.setDescription(
+			"Manage leaderboard that sorts server members by characters sent"
+		);
+	cmd.addSubcommand(
+		new Builders.SlashCommandSubcommandBuilder()
+			.setName("send")
+			.setDescription("Calculate and send a new leaderboard")
+			.addStringOption(
+				new Builders.SlashCommandStringOption()
+					.setName("time-period")
+					.setDescription(
+						"How recent do messages have to be to be considered; defaults to all"
+					)
+					.setRequired(false)
+					.addChoice("All messages", "all")
+					.addChoice("Last 30 days", "month")
+					.addChoice("Last 7 days", "week")
+			)
+	);
+	return [
+		{
+			builder: cmd,
+			handler: handleCommand,
+		},
+	];
+}
+
+export async function handleCommand(
+	interaction: Discord.CommandInteraction
+): Promise<void> {
+	switch (interaction.options.getSubcommand()) {
+		case "send": {
+			try {
+				if (!interaction.channel) {
+					throw new Error("Channel must exist");
+				}
+				// TODO: if one is already being sent, don't allow another user to run command
+				const period =
+					interaction.options.getString("time-period", false) ??
+					"all";
+				await interaction.editReply(
+					`Will send a leaderboard (period: ${period}) to ${interaction.channel}, but calculations are necessary.\nThis may take a while...`
+				);
+
+				const [delta, [channels, msgs, failed, msgId]] =
+					(await utils.timeFunction(
+						async () =>
+							await sendLeaderboard(
+								interaction.channel as
+									| Discord.TextChannel
+									| Discord.ThreadChannel,
+								period
+							)
+					)) as [
+						number,
+						utils.ThenArg<ReturnType<typeof sendLeaderboard>>
+					];
+				await interaction.editReply(`✅ Sent [here](https://discord.com/channels/${
+					interaction.guildId as string
+				}/${interaction.channelId as string}/${msgId})
+
+Took ${delta}ms, combed through ${channels} channels and ${msgs} messages.
+${
+	failed.length
+		? "❌ Failed to go through the following channels: " +
+		  failed.map((c) => "<#" + c.id + ">").join(", ")
+		: "Did not fail to go through any channel"
+}`);
+				// TODO: if any failed, say I can't cache it (whenever I do start caching it with checkpoints)
+			} catch (e) {
+				await interaction.editReply("❌ Something went wrong.");
+			}
```

Comment:
I think Discord only allows interaction replies to be edited after 30 minutes.
Additional error handling may be necessary here (either a try/catch or a `.catch`).

Editing the reply inside the catch block will throw an error up the stack and may even kill the bot